ecl/hthor/hthor.cpp

```diff
@@ -1641,28 +1656,42 @@ class CHThorPipeThroughActivity : public CHThorSimpleActivityBase, implements IP
 private:
     bool waitForPipe()
     {
-        if (firstRead)
+        Owned<IPipeProcessException> pipeException;
+        try
         {
-            pipeOpened.wait();
-            firstRead = false;
+            if (firstRead)
+            {
+                pipeOpened.wait();
+                firstRead = false;
+            }
+            if (!pipe)
+                return false;  // done
+            if (!readTransformer->eos())
+                return true;
+        }
+        catch (IPipeProcessException *e)
+        {
+            pipeException.setown(e);
         }
-        if (!pipe)
-            return false;  // done
-        if (!readTransformer->eos())
-            return true;
         verifyPipe();
+        if (pipeException) // NB: verifyPipe may throw error based on pipe prog. output 1st.
+            throw pipeException.getClear();
         if (recreate && !inputExhausted)
             pipeOpened.wait();
         return false;
     }
-
     void openPipe(char const * cmd)
     {
         pipeCommand.setown(cmd);
         pipe.setown(createPipeProcess(agent.queryAllowedPipePrograms()));
         if(!pipe->run(NULL, cmd, ".", true, true, true, 0x10000))
-            throw MakeStringException(2, "Could not run pipe process %s", cmd);
-        writeTransformer->writeHeader(pipe);
+        {
+            // NB: pipe->run can't rely on the child process failing fast enough to return false here, failure picked up later with stderr context.
+            WARNLOG(2, "Could not run pipe process %s", cmd);
+        }
+        else
+            writeTransformer->writeHeader(pipe);
```

Comment:
If we dont try/catch around this for an IPipeProcessException, who will catch it ?

Reply:
it will propagate up as with any other exception, ultimately if nothing else wants to handle it, the framework (in EclAgent) will catch, handle and report it.